arrows/vxl/morphology.cxx

```diff
@@ -0,0 +1,325 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "morphology.h"
+
+#include <arrows/vxl/image_container.h>
+
+#include <vital/util/enum_converter.h>
+
+#include <vital/range/iota.h>
+
+#include <vil/algo/vil_binary_closing.h>
+#include <vil/algo/vil_binary_dilate.h>
+#include <vil/algo/vil_binary_erode.h>
+#include <vil/algo/vil_binary_opening.h>
+#include <vil/algo/vil_structuring_element.h>
+#include <vil/vil_plane.h>
+#include <vil/vil_transform.h>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+namespace {
+
+enum morphology_mode
+{
+  MORPHOLOGY_erode,
+  MORPHOLOGY_dilate,
+  MORPHOLOGY_open,
+  MORPHOLOGY_close,
+  MORPHOLOGY_none,
+};
+
+ENUM_CONVERTER( morphology_converter, morphology_mode,
+                { "erode", MORPHOLOGY_erode }, { "dilate", MORPHOLOGY_dilate },
+                { "open", MORPHOLOGY_open }, { "close", MORPHOLOGY_close },
+                { "none", MORPHOLOGY_none } );
+
+enum element_mode
+{
+  ELEMENT_disk,
+  ELEMENT_jline,
+  ELEMENT_iline,
+};
+
+ENUM_CONVERTER( element_converter, element_mode, { "disk", ELEMENT_disk },
+                { "iline", ELEMENT_iline }, { "jline", ELEMENT_jline } );
+
+enum combine_mode
+{
+  COMBINE_none,
+  COMBINE_union,
+  COMBINE_intersection,
+};
+
+ENUM_CONVERTER( combine_converter, combine_mode, { "none", COMBINE_none },
+                { "union", COMBINE_union },
+                { "intersection", COMBINE_intersection } );
+
+inline bool
+union_functor( bool x1, bool x2 )
+{
+  return x1 || x2;
+}
+
+inline bool
+intersection_functor( bool x1, bool x2 )
+{
+  return x1 && x2;
+}
+
+} // namespace
+
+// ----------------------------------------------------------------------------
+class morphology::priv
+{
+public:
+  priv()
+  {
+  }
+
+  ~priv();
+  // Setup structuring elements
+  void setup_internals();
+  // Compute the morpholigical operation on a single plane of an input image
+  void
+  apply_morphology_to_plane( vil_image_view< bool > const& input,
+                             vil_image_view< bool >& output,
+                             unsigned plane_index );
+  // Perform a morphological operation and optionally combine across channels
+  vil_image_view< bool >
+  perform_morphological_operations( vil_image_view< bool > const& input );
+
+  bool setup{ false };
+  vil_structuring_element morphological_element;
+
+  morphology_mode morphology_type{ MORPHOLOGY_dilate };
+  element_mode element_type{ ELEMENT_disk };
+  combine_mode combine_type{ COMBINE_none };
+  double element_size{ 1.0 };
```

Comment:
I think something bigger than 1.0 would be a better default.  According to the docs, a size of 1.0 will make a no-op filter.  For a disk, pixels are included in the mask if `x^x+y^y<r^r` where `r` is element size and `(x,y) == (0,0)` is the center pixel of the mask.  So 1.0 will get you just the center pixel, which doesn't do anything.  Values between 1.0 and sqrt(2) would give you a 3x3 `+` shape.  Values between sqrt(2) and 2.0 would give you a 3x3 box.

I would just go with 1.5.  That will give you the 3x3 box, which is a pretty common default for morphology.

https://vxl.github.io/doc/release/core/vil/html/vil__structuring__element_8cxx_source.html

Reply:
I agree that this makes sense and have updated it to 1.5. However, this is a bit confusing because the default burnout configs use 1.0 in several places. I wonder if that ends up being a no-op or if perhaps the documentation is wrong, since it incorrectly states that each element is being raised to the power of itself, e.g. `x^x`.

Reply:
It's funny that I read that equation and didn't even notice the mistake,  I suppose because I know what it was meant to be.  The code is indeed `x*x+y*y<r*r` which could be documented as `x^2+y^2<r^2`.  It's possible that there is some numerical precision condition where `int(1) < double(1.0 * 1.0) == true`, but I haven't check that.  It's safer to go with something explicitly not on the boundary condition.

I wonder if the fact that it is 1.0 on burnout config indicates that @mattdawkins was turning off the morphology?  It would seem like that's not the best way to do it.  Or maybe still using the combination over planes aspect of this code without the erosion or dilation?

Reply:
Would have to revisit the code, we definitely used morphology in at least one place but it was usually just strict dilations of final masks, no closing or opening ops

Reply:
@mattdawkins I've been slightly confused about the convention for the masks. Are true boolean values indicating the location of burnin?

The thresholding operations seemed to suggest that was the case, but the filtering operations seemed to suggest the opposite was true.

Reply:
Which filtering options specifically? I'll send you a link on chat to file where most of the morphology I remember was happening.